server/src/main/java/io/deephaven/server/console/ConsoleServiceGrpcImpl.java

```diff
@@ -303,57 +316,145 @@ public void onCompleted() {
-            logBuffer.unsubscribe(this);
+        public void start() {
+            logBuffer.subscribe(this);
+            scheduler.runImmediately(this);
         }
 
+        public void stop() {
+            GrpcUtil.safelyComplete(client);
+        }
+
+        // ------------------------------------------------------------------------------------------------------------
+
         @Override
         public void record(LogBufferRecord record) {
+            if (done) {
+                return;
+            }
             // only pass levels the client wants
             if (request.getLevelsCount() != 0 && !request.getLevelsList().contains(record.getLevel().getName())) {
                 return;
             }
-
-            // since the subscribe() method auto-replays all existing logs, filter to just once this
-            // consumer probably hasn't seen
+            // since the subscribe() method auto-replays all existing logs, filter to just once this consumer probably
+            // hasn't seen
             if (record.getTimestampMicros() < request.getLastSeenLogTimestamp()) {
                 return;
             }
+            // Note: we can't send record off-thread without doing a deepCopy.
+            // io.deephaven.io.logger.LogBufferInterceptor owns io.deephaven.io.logger.LogBufferRecord.getData.
+            // We can side-step this problem by creating the appropriate response here.
+            final LogSubscriptionData payload = LogSubscriptionData.newBuilder()
+                    .setMicros(record.getTimestampMicros())
+                    .setLogLevel(record.getLevel().getName())
+                    .setMessage(record.getDataString())
+                    .build();
+            enqueue(payload);
+        }
+
+        // ------------------------------------------------------------------------------------------------------------
+
+        @Override
+        public void run() {
+            while (true) {
+                if (!guard.compareAndSet(false, true)) {
+                    return;
+                }
+                boolean queueIsKnownEmpty;
+                try {
+                    queueIsKnownEmpty = sendInternal();
+                } finally {
+                    guard.set(false);
+                }
+                if (done) {
+                    return;
+                }
+                if (!client.isReady()) {
+                    // When !client.isReady(), onReady() is going to be called and will handle the reschedule.
+                    // Note: it's important that client.isReady() is checked _outside_ of the guard block (otherwise,
+                    // the onReady() call could execute and schedule before the current thread exits the guard block,
+                    // and we'd fail to successfully reschedule).
+                    return;
+                }
+                if (queueIsKnownEmpty) {
+                    scheduler.runAfterDelay(SUBSCRIBE_TO_LOGS_SEND_MILLIS, this);
```

Comment:
This makes me wonder if we should have a repeat scheduling option in the scheduler.

Reply:
I've tried to plumb it, but ran into limitations about how we shim the scheduler. Added a note.